elastic/kibana
# [Actions] Update `relatedSavedObject` to accept array of `space_ids`

Resolves https://github.com/elastic/kibana/issues/144112

## Summary

Updates typing for `RelatedSavedObjects` from `namespace?: string` to `space_ids: string`. Previously we allowed a single namespace value for a related saved object and having an undefined value for `namespace` implied the default space. This updated typing allows multiple space IDs (to support shared saved objects) and a saved object in the default space must specify the `default` space ID.

This data is persisted in the `action_task_param` saved object, which is then read into the action executor and then persisted in the event log.
* Updated event log mapping to accept `space_ids` field in the `kibana.saved_objects` array. I kept the previous mapping for `namespace` to support backwards compatibility
* Added migration for `action_task_param` saved object.

Also had to update the event log query to perform an OR query against both the old and new field.

## To Verify

**Verify event log docs for action execution show `space_ids` array**
1. Create a rule that fires an action in the default space. Let it run and inspect the event log `execute` documents for the rule and the action. Both should have a `kibana.saved_objects` array with `space_ids: ["default"]` in each array object.
2. Create a rule that fires an action from a preconfigured connector in the default space. Let it run and inspect the event log `execute` documents for the rule and the action. For the rule, the `kibana.saved_objects` array should have 1 entry for the rule saved object and the default space id. For the action, the `kibana.saved_objects` array should have 1 entry for the rule saved object and the default space id and 1 entry for the preconfigured connector and no space_id (because preconfigured connectors do not belong to a space.
3. Repeat 1 & 2 with a rule in a custom space. Verify that the space IDs in the event log doc accurately reflects the custom space id.

**Verify that event log queries are backwards compatible**
1. On `main`, create a rule and let it run a few times.
2. Switch to this PR branch and let it run a few more times.
3. Verify that on the rule details page execution history, you can see information about all runs.
4. Repeat steps using rule in a custom space

**Verify the migration**
1. On an older branch like 8.5, exclude actions from running by adding `xpack.task_manager.internal.exclude_task_types: ['actions:*']` to your Kibana config.
2. Create a rule in the default space and one in the custom space that will trigger actions. Verify that you see `action_task_params` saved objects in the `.kibana` index.
3. Run this PR branch against the same data. Make sure you still have `xpack.task_manager.internal.exclude_task_types: ['actions:*']` in your config.
4. Verify that the `action_task_params` in the `.kibana` index have been migrated correctly.
5. Remove the `xpack.task_manager.internal.exclude_task_types` config and verify that the actions are picked up and run successfully.
